Druid/DRCogitCodeGenerator.class.st

```diff
@@ -311,6 +320,31 @@ DRCogitCodeGenerator >> generateJump: jumpKind target: aTarget [
 	
 ]
 
+{ #category : #'code-generation' }
+DRCogitCodeGenerator >> generateLoop: aDRLoop [
+
+	| innerGenerator jitCondition loopPath isTrue |
+	loopPath := aDRLoop bodyBlocks.
+
+	self assert:
+		(loopPath allSatisfy: [ :b | b isDominatedBy: aDRLoop headerBlock ]).
+	blocks removeAll: loopPath.
+	innerGenerator := self innerCompileBlocks: loopPath.
```

Comment:
Ok, here the variable name confused me at first.
This is not a loop path! This is all the blocks dominated by the loop header!

We could use the same for the ifs without having traces :')

Reply:
It is not **all** the blocks dominated by the header, it is the _loopBody_ :P (or `loopBodyBlocks` to say that they could be many).
The assertion is to be safe in case of jumps to the middle of the body (I'm not sure what happens then).

The if compilation also does this to know if blocks are shared between the staged and not staged paths.

The traces are there to avoid some phi functions. I don't remember the problem but we can check that easily.

Reply:
yes, it's not the ones dominated by the header. It's the ones dominated by the loop body block.
What we need is all the blocks up to the domination frontier.

Yes let's check :)